File: blinky-serial-mcp/src/lib/audio-lock.ts
/**
 * Audio lock — mutual exclusion for audio playback.
 *
 * All devices share the same room, so only one process may play audio at a
 * time. This module implements a file-based lock using atomic O_CREAT|O_EXCL
 * creation, matching the protocol in ml-training/tools/ab_test_multidev.cjs.
 *
 * Lock file: /tmp/blinky-audio.lock
 * Contents:  { pid, ports, started } (JSON)
 */

import * as fs from 'node:fs';

const LOCK_PATH = '/tmp/blinky-audio.lock';

let lockHeld = false;

interface LockInfo {
  pid: number;
  ports: string[];
  started: string;
}

/**
 * Attempt to acquire the audio lock.
 *
 * Uses atomic file creation (O_CREAT|O_EXCL) so two processes racing will
 * never both succeed. If a stale lock is detected (owning PID no longer
 * exists), it is removed and acquisition is retried once.
 *
 * Returns true if the lock was acquired, false if another live process holds it.
 */
export function acquireAudioLock(ports: string[]): boolean {
  try {
    const fd = fs.openSync(
      LOCK_PATH,
      fs.constants.O_CREAT | fs.constants.O_EXCL | fs.constants.O_WRONLY,
    );
    const info: LockInfo = {
      pid: process.pid,
      ports,
      started: new Date().toISOString(),
    };
    fs.writeSync(fd, JSON.stringify(info));
    fs.closeSync(fd);
    lockHeld = true;
    return true;
  } catch (err: unknown) {
    const e = err as NodeJS.ErrnoException;
    if (e.code === 'EEXIST') {
      // Lock file already exists — check if the holder is still alive.
      try {
        const raw = fs.readFileSync(LOCK_PATH, 'utf-8');
        const info: LockInfo = JSON.parse(raw);

        try {
          // Signal 0 checks existence without actually signaling.
          process.kill(info.pid, 0);
        } catch (killErr: unknown) {
          if ((killErr as NodeJS.ErrnoException).code === 'ESRCH') {
            // Process is dead — stale lock. Remove and retry once.
            try {
              fs.unlinkSync(LOCK_PATH);
            } catch {
              // Another process may have grabbed it between our read and unlink.
            }
            return acquireAudioLock(ports);
          }
        }

        // Process is alive — lock is legitimately held.
        console.error(
          `Audio lock held by PID ${info.pid} on ${info.ports.join(', ')} (started ${info.started})`,
        );
        console.error(
          'All devices share the same room — concurrent audio tests are invalid.',
        );
        console.error(
          `Remove ${LOCK_PATH} manually if the process is stuck.`,
        );
      } catch {
        // Could not read/parse the lock file.
        console.error(
          `Audio lock exists at ${LOCK_PATH}. Another test may be running.`,
        );
      }
      return false;
    }
    // Unexpected error (permissions, disk full, etc.) — propagate.
    throw e;
  }
}

/**
 * Release the audio lock. Safe to call even if the lock is not held.
 */
export function releaseAudioLock(): void {
  if (!lockHeld) return;
  try {
    fs.unlinkSync(LOCK_PATH);
  } catch {
    // Ignore — file may already be gone.
  }
  lockHeld = false;
}

/**
 * Check whether the audio lock is currently held by a live process.
 *
 * If a stale lock is found (dead PID), it is cleaned up automatically and
 * the function reports unlocked.
 */
export function isAudioLocked(): {
  locked: boolean;
  holder?: LockInfo;
} {
  try {
    const raw = fs.readFileSync(LOCK_PATH, 'utf-8');
    const info: LockInfo = JSON.parse(raw);

    try {
      process.kill(info.pid, 0);
    } catch (killErr: unknown) {
      if ((killErr as NodeJS.ErrnoException).code === 'ESRCH') {
        // Stale lock — clean it up.
        try {
          fs.unlinkSync(LOCK_PATH);
        } catch {
          // Race with another cleaner — fine.
        }
        return { locked: false };
      }
    }

    return { locked: true, holder: info };
  } catch {
    // No lock file or unreadable — not locked.
    return { locked: false };
  }
}

// --- Process exit cleanup ---
// Release the lock on any exit path so we never leave stale locks behind.

function cleanup() {
  releaseAudioLock();
}

process.on('exit', cleanup);
process.on('SIGINT', () => {
  cleanup();
  process.exit(128 + 2);
});
process.on('SIGTERM', () => {
  cleanup();
  process.exit(128 + 15);
});
process.on('uncaughtException', (err) => {
  console.error('Uncaught exception — releasing audio lock:', err);
  cleanup();
  process.exit(1);
});

File: blinky-server/blinky_server/testing/audio_lock.py
"""Audio lock — mutual exclusion for audio playback.

All devices share the same room, so only one process may play audio at a time.
File-based lock using atomic O_CREAT|O_EXCL, compatible with the TypeScript
and CJS implementations at /tmp/blinky-audio.lock.
"""

from __future__ import annotations

import atexit
import errno
import json
import logging
import os
import signal
from types import FrameType
from typing import Any

log = logging.getLogger(__name__)

LOCK_PATH = "/tmp/blinky-audio.lock"
_lock_held = False


def acquire_audio_lock(device_ids: list[str] | None = None) -> bool:
    """Acquire the audio lock. Returns True on success, False if held by another live process."""
    global _lock_held
    try:
        fd = os.open(LOCK_PATH, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o644)
        info = {
            "pid": os.getpid(),
            "ports": device_ids or [],
            "started": __import__("datetime").datetime.now().isoformat(),
        }
        os.write(fd, json.dumps(info).encode())
        os.close(fd)
        _lock_held = True
        return True
    except FileExistsError:
        # Lock file exists — check if holder is alive
        try:
            with open(LOCK_PATH) as f:
                info = json.load(f)
            try:
                os.kill(info["pid"], 0)  # Check existence
            except OSError as e:
                if e.errno == errno.ESRCH:
                    # No such process — stale lock, remove and retry once
                    try:
                        os.unlink(LOCK_PATH)
                    except OSError:
                        pass
                    return acquire_audio_lock(device_ids)
                # EPERM = process exists but owned by another user — lock is valid
            # Process alive — lock legitimately held
            log.warning(
                "Audio lock held by PID %d on %s (started %s)",
                info["pid"],
                info.get("ports", []),
                info.get("started", "?"),
            )
            return False
        except Exception:
            log.warning("Audio lock exists at %s but could not read it", LOCK_PATH)
            return False


def release_audio_lock() -> None:
    """Release the audio lock. Safe to call even if not held."""
    global _lock_held
    if not _lock_held:
        return
    try:
        os.unlink(LOCK_PATH)
    except OSError:
        pass
    _lock_held = False


def is_audio_locked() -> tuple[bool, dict[str, Any] | None]:
    """Check if the audio lock is held. Cleans up stale locks."""
    try:
        with open(LOCK_PATH) as f:
            info = json.load(f)
        try:
            os.kill(info["pid"], 0)
        except OSError as e:
            if e.errno == errno.ESRCH:
                # Stale lock — dead process
                try:
                    os.unlink(LOCK_PATH)
                except OSError:
                    pass
                return False, None
            # EPERM = process exists, lock is valid
        return True, info
    except (FileNotFoundError, json.JSONDecodeError, KeyError):
        return False, None


def _cleanup() -> None:
    release_audio_lock()


atexit.register(_cleanup)

# Signal handlers — release lock on SIGINT/SIGTERM
_orig_sigint = signal.getsignal(signal.SIGINT)
_orig_sigterm = signal.getsignal(signal.SIGTERM)


def _sigint_handler(signum: int, frame: FrameType | None) -> None:
    _cleanup()
    if callable(_orig_sigint):
        _orig_sigint(signum, frame)
    else:
        raise KeyboardInterrupt


def _sigterm_handler(signum: int, frame: FrameType | None) -> None:
    _cleanup()
    raise SystemExit(128 + 15)


signal.signal(signal.SIGINT, _sigint_handler)
signal.signal(signal.SIGTERM, _sigterm_handler)
